File: node-bindings/classic-node/__test__/update_decisions_conformance.ts
import * as classic from "../index.js";

type JsonObject = Record<string, any>;

/** Project only native strict-version failures, allowing unrelated errors to fail execution. */
function observe(call: () => boolean): JsonObject {
    try {
        return {hasUpdate: call(), error: null};
    } catch (failure) {
        if (!(failure instanceof Error) || !failure.message.startsWith("Version error:")) throw failure;
        return {hasUpdate: null, error: {code: "invalid_version"}};
    }
}

/** Compare explicit versions through class and free-function bindings without fetching. */
export function observeUpdateDecisions(fixture: JsonObject): JsonObject {
    if (Object.keys(fixture).sort().join() !== "current,latest"
        || typeof fixture.current !== "string" || typeof fixture.latest !== "string") {
        throw new Error("unsupported update decision fixture");
    }
    const client = new classic.GithubClient("conformance", "unused");
    const result = observe(() => client.hasUpdate(fixture.current, fixture.latest));
    const standalone = observe(() => classic.hasUpdate(fixture.current, fixture.latest));
    if (JSON.stringify(result) !== JSON.stringify(standalone)) throw new Error("update binding entrypoints disagree");
    return result;
}

File: python-bindings/tests/update_decisions_conformance.py
"""Observe strict version decisions through actual Python update bindings."""

from collections.abc import Mapping
from typing import Any


def observe_update_decisions(fixture: Mapping[str, Any]) -> dict[str, Any]:
    """Compare explicit versions without invoking any network-facing client method."""
    import classic_update

    if set(fixture) != {"current", "latest"} or any(
            not isinstance(value, str) for value in fixture.values()
    ):
        raise ValueError("unsupported update decision fixture")
    client = classic_update.GithubClient("conformance", "unused")
    try:
        result = client.has_update(fixture["current"], fixture["latest"])
        return {"hasUpdate": result, "error": None}
    except ValueError as failure:
        # Only strict semver errors belong to this domain; initialization failures propagate.
        if not str(failure).startswith("Version error: Version error:"):
            raise
        return {"hasUpdate": None, "error": {"code": "invalid_version"}}
